src/Stack/MetricsCollector.php

```diff
@@ -0,0 +1,229 @@
+    }
+
+    public function collect_request_metrics()
+    {
+        $request_type = $this::_get_request_type();
+        $display      = 0;
+        $precision    = 12;
+        $request_time = timer_stop($display, $precision);
+        $peak_memory  = memory_get_peak_usage();
+
+        $this->metrics->get_counter('wp.requests')->incBy(1, [$request_type]);
+        $this->metrics->get_histogram('wp.peak_memory')->observe($peak_memory, [$request_type]);
+        $this->metrics->get_histogram('wp.page_generation_time')->observe($request_time, [$request_type]);
+
+        if ($this::_do_collect_wpdb_metrics()) {
+            $this->metrics->get_histogram('wpdb.query_time')->observe($this->wpdb_stats['query_time'], [$request_type]);
+            $this->metrics->get_histogram('wpdb.num_queries')->observe($this->wpdb_stats['num_queries'], [$request_type]);
+            $this->metrics->get_histogram('wpdb.num_slow_queries')->observe($this->wpdb_stats['num_slow_queries'], [$request_type]);
+        }
+    }
+
+    public function register_endpoint()
+    {
+        $version   = '1';
+        $namespace = 'stack/v' . $version;
+        $base      = 'metrics';
+
+        register_rest_route($namespace, '/' . $base, array(
+            array(
+                'methods'  => \WP_REST_Server::READABLE,
+                'callback' => [$this->metrics, 'render']
+            )
+        ));
+    }
+
+    public function collect_wpdb_stats($query_data, $query, $query_time, $query_callstack, $query_start)
+    {
+        if (!$this::_do_collect_wpdb_metrics()) {
+            return;
+        }
+
+        $query_duration = $query_time - $query_start;
+        $this->wpdb_stats['num_queries'] += 1;
+        $this->wpdb_stats['query_time'] += $query_time;
+
+        if ($query_time > $this->wpdb_stats['slow_query_treshold']) {
+            $this->wpdb_stats['num_slow_queries'] += 1;
+        }
+
+        return $query_data;
+    }
+
+    public function track_woocomerce_order($order_id, $old_status, $new_status)
+    {
+        if ($new_status == 'completed') {
+            $this->metrics['woocommerce.orders']->inc();
+        }
+    }
+
+    public function track_woocomerce_checkout()
+    {
+        $this->metrics['woocommerce.checkouts']->inc();
+    }
+
+    private function register_hooks()
+    {
+        add_action('rest_api_init', [$this, 'register_endpoint']);
+        add_action('shutdown', [$this, 'collect_request_metrics']);
+
+        if ($this::_do_collect_wpdb_metrics()) {
+            add_filter('log_query_custom_data', [$this, 'collect_wpdb_stats'], 10, 5);
+        }
+
+        if ($this::_do_collect_woocommerce_metrics()) {
+            add_action('woocommerce_checkout_billing', [$this, 'track_woocomerce_checkout']);
+            add_action('woocommerce_order_status_changed', [$this, 'track_woocomerce_order'], 10, 3);
+        }
+    }
+
+    private function _do_collect_wpdb_metrics()
+    {
+        return isset($this->wpdb_stats);
```

Comment:
Is a little misleading since it has a level of indirection. Maybe we can use a constant? from env?
Same for `_do_collect_woocommerce_metrics `

Reply:
This is not meant to be a flag to enable / disable the collectors, but a guard against trying to collect when there is no support for that, and it's meant to abstract that.  
It's more like `is woocommerce installed?` and `are db queries logged?`.  
I do need to fix the `wpdb` guard, though, since the `log_query_custom_data` is only called if `SAVEQUERIES` is enabled, so I should check for that constant.

Reply:
Since it's called in multiple places around here, I think it's best to keep it in a method instead of duplicating the logic in it several times.

Reply:
The name should be without the leading `_` and camelCased, though.

Reply:
It makes sense. `canCollectWPDBMetrics` can also be an option.